Add torch.logit function

xiaomengy opened the pull request:
Summary: Add torch.logit function

Test Plan: buck test mode/dev-nosan //caffe2/test:torch -- "logit"

Differential Revision: D22406912

BIT-silence commented:
link https://github.com/pytorch/pytorch/issues/37060

mruberry commented:
Thanks for putting a PR up, @BIT-silence. I'll get you a review ASAP. Want to add a couple helpful links now before I lose them:

- SciPy's logit function: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.logit.html
- SciPy implementation: https://github.com/scipy/scipy/blob/8dba340293fe20e62e173bdf2c10ae208286692f/scipy/special/_logit.c.src#L23.

ngimel reviewed (approved):
Small comment about type of scalar arg

mruberry reviewed (changes requested):
Hey @BIT-silence! What's here looks pretty good, but there are a few more things to do:

- add a doc entry in docs/source/tensors.rst and docs/source/torch.rst
- add the actual docs in torch/_tensor_docs.py and torch/_torch_docs.py
- test the backward works as expected by adding an entry to torch/testing/_internal/common_method_invocations.py
- add an entry in torch/_overrides.py so people can override this function when using the torch_function interface

Also, what do you think about adding an explicit test for extremal values in test_torch? Like -inf, -5, 0, 1, 2, inf, NaN?

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2418,6 +2418,19 @@
 
 - func: sigmoid.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
 
+- func: logit(Tensor self, float? eps=0.0) -> Tensor
```
why make it optional and give it a non-optional default value?  AFAICT, this just lets you write `torch.logit(self, eps=None)`, but why would I want to do that?

BIT-silence replied:
Done.

ngimel commented:
> Also, what do you think about adding an explicit test for extremal values in test_torch? Like -inf, -5, 0, 1, 2, inf, NaN?

Instead of adding separate tests for extremal values, perhaps it makes sense to add those to TorchMathTest so that all functions there are tests? It already has "interesting" finite values, but it indeed misses infs and nans.

BIT-silence commented:
> > Also, what do you think about adding an explicit test for extremal values in test_torch? Like -inf, -5, 0, 1, 2, inf, NaN?
> 
> Instead of adding separate tests for extremal values, perhaps it makes sense to add those to TorchMathTest so that all functions there are tests? It already has "interesting" finite values, but it indeed misses infs and nans.

I think we can consider open another issue for this. I tried to add inf and nan in the current tests, but some of the existing test will fail because of the different behavior for outputing inf or nan. If we want to make everything match with numpy/scipy, maybe some more works are needed.

BIT-silence commented on aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu:
```diff
@@ -19,6 +22,26 @@ void sigmoid_backward_kernel_cuda(TensorIterator& iter) {
+      [&]() {
+        AT_SKIP_BFLOAT16_IF_NOT_ROCM(scalar_t, "logit_cuda", [&] {
+          using T_ACC = acc_type<scalar_t, true>;
+          const scalar_t eps = eps_scalar.to<scalar_t>();
```
Done.

mruberry commented:
> > > Also, what do you think about adding an explicit test for extremal values in test_torch? Like -inf, -5, 0, 1, 2, inf, NaN?
> > 
> > 
> > Instead of adding separate tests for extremal values, perhaps it makes sense to add those to TorchMathTest so that all functions there are tests? It already has "interesting" finite values, but it indeed misses infs and nans.
> 
> I think we can consider open another issue for this. I tried to add inf and nan in the current tests, but some of the existing test will fail because of the different behavior for outputing inf or nan. If we want to make everything match with numpy/scipy, maybe some more works are needed.

Can you elaborate on this? It's important we be compatible with NumPy or we may have to go through a painful deprecation process later. We already test several torch functions against NumPy for extremal values like this and comparing their values directly hasn't been a problem so far.

BIT-silence commented:
> > > > Also, what do you think about adding an explicit test for extremal values in test_torch? Like -inf, -5, 0, 1, 2, inf, NaN?
> > > 
> > > 
> > > Instead of adding separate tests for extremal values, perhaps it makes sense to add those to TorchMathTest so that all functions there are tests? It already has "interesting" finite values, but it indeed misses infs and nans.
> > 
> > 
> > I think we can consider open another issue for this. I tried to add inf and nan in the current tests, but some of the existing test will fail because of the different behavior for outputing inf or nan. If we want to make everything match with numpy/scipy, maybe some more works are needed.
> 
> Can you elaborate on this? It's important we be compatible with NumPy or we may have to go through a painful deprecation process later. We already test several torch functions against NumPy for extremal values like this and comparing their values directly hasn't been a problem so far.

I think it's better to let me do that in another PR and let this focus on adding logit function. Do you think it make sense?

mruberry commented:
> > > > > Also, what do you think about adding an explicit test for extremal values in test_torch? Like -inf, -5, 0, 1, 2, inf, NaN?
> > > > 
> > > > 
> > > > Instead of adding separate tests for extremal values, perhaps it makes sense to add those to TorchMathTest so that all functions there are tests? It already has "interesting" finite values, but it indeed misses infs and nans.
> > > 
> > > 
> > > I think we can consider open another issue for this. I tried to add inf and nan in the current tests, but some of the existing test will fail because of the different behavior for outputing inf or nan. If we want to make everything match with numpy/scipy, maybe some more works are needed.
> > 
> > 
> > Can you elaborate on this? It's important we be compatible with NumPy or we may have to go through a painful deprecation process later. We already test several torch functions against NumPy for extremal values like this and comparing their values directly hasn't been a problem so far.
> 
> I think it's better to let me do that in another PR and let this focus on adding logit function. Do you think it make sense?

It might, but I don't really understand why it should be a separate PR. Are you saying the function is computing the right values but the test suite would need changes to detect this?

ngimel commented:
>Can you elaborate on this? It's important we be compatible with NumPy or we may have to go through a painful deprecation process later. We already test several torch functions against NumPy for extremal values like this and comparing their values directly hasn't been a problem so far.

Actually, it has been a problem, most recent example being exp for extremal complex values. Polygamma/digamma also don't compare well with their scipy counterparts. 
If logit is actually compatible with scipy including nan/inf values, I suggest a compromise might be to run it in TorchMathTest with a special input that includes infs/nans (TorchMathMeta allows one to specify custom input generation function), and leave the rest of the functions in TorchMathTest for later.

mruberry commented:
> > Can you elaborate on this? It's important we be compatible with NumPy or we may have to go through a painful deprecation process later. We already test several torch functions against NumPy for extremal values like this and comparing their values directly hasn't been a problem so far.
> 
> Actually, it has been a problem, most recent example being exp for extremal complex values. Polygamma/digamma also don't compare well with their scipy counterparts.
> If logit is actually compatible with scipy including nan/inf values, I suggest a compromise might be to run it in TorchMathTest with a special input that includes infs/nans (TorchMathMeta allows one to specify custom input generation function), and leave the rest of the functions in TorchMathTest for later.

I think we're all talking past each other. Yes, trying to add values to a suite of tests has been an issue. But comparing a particular function against NumPy with extremal values has never been an issue.

BIT-silence commented:
> > Can you elaborate on this? It's important we be compatible with NumPy or we may have to go through a painful deprecation process later. We already test several torch functions against NumPy for extremal values like this and comparing their values directly hasn't been a problem so far.
> 
> Actually, it has been a problem, most recent example being exp for extremal complex values. Polygamma/digamma also don't compare well with their scipy counterparts.
> If logit is actually compatible with scipy including nan/inf values, I suggest a compromise might be to run it in TorchMathTest with a special input that includes infs/nans (TorchMathMeta allows one to specify custom input generation function), and leave the rest of the functions in TorchMathTest for later.

For logit itself, there is another issue. Here I added one argument which is eps so the input is clamped to [eps, 1-eps] so that the user can remove the inf/nan in the model. So actually torch.logit is clamp and logit. Then for the input values out of [0, 1], the torch.logit(x) will output inf while the scipy.special.logit will output nan. Considering the actual usage in the models, I think the eps argument is useful in most cases. Another reason I added this eps is from the gradient reason, to do clamp first will make the gradient well defined for values out of [0, 1]. Here the question is how we define the of the output matches scipy. If we define the function as clamp to [eps, 1 - eps] then logit, the current logit function already matches. Otherwise we can also consider to let logit do exactly scipy.special.logit does when eps == None, and if there is eps, we do clamp then logit. I'm not sure what do you think about this.

ngimel commented:
I think your proposal makes sense, there should be a way to recover scipy behavior, and preferably it should be default. So torch.logit(x) should produce `nan` for values outside of [0,1], torch.logit(x, eps=0) should produce `-+inf` for values outside of [0,1]. 
I've updated TestTorchMathOps so it now tests for nan and +-inf values also, so there's no need for special test. The exact comparisons could be shadowed by `replace_inf_with_nan` flag though.

BIT-silence commented:
> I think your proposal makes sense, there should be a way to recover scipy behavior, and preferably it should be default. So torch.logit(x) should produce `nan` for values outside of [0,1], torch.logit(x, eps=0) should produce `-+inf` for values outside of [0,1].
> I've updated TestTorchMathOps so it now tests for nan and +-inf values also, so there's no need for special test. The exact comparisons could be shadowed by `replace_inf_with_nan` flag though.

I have updated the impl to output exact same thing as scipy when eps is None and remove the 'replace_inf_with_nan' flag in test. In this case, the gradient for values out of [0, 1] is also nan.

ngimel commented on torch/_torch_docs.py:
```diff
@@ -5444,6 +5444,36 @@ def merge_dicts(*dicts):
+logit(input, eps=None, out=None) -> Tensor
+
+Returns a new tensor with the logit of the elements of :attr:`input`.
+:attr:`input` is clamped to [eps, 1 - eps] when eps is not None.
```
Can you please note that when eps is None then logit(p) for p<0 or p>1 yields nan.

BIT-silence replied:
Done.

ngimel commented on torch/csrc/jit/runtime/symbolic_script.cpp:
```diff
@@ -787,6 +787,23 @@ const std::vector<std::string> functions = {
 
             return result, backward
 
+        def logit(self, eps: Optional[float]):
```
why are you defining symbolic gradient? This would cause scripted module run this implementation instead of built-in backward, and without fuser this will be much slower. And I'm not sure if fuser (which of them?) is able to efficiently fuse this backward.

BIT-silence replied:
I don't know why if I didn't add it here, there will be an error that autodiff node number didn't match. So I added it here to test that.

ngimel replied:
Which test was failing?

ngimel reviewed (approved):
Looks good now, modulo a small documentation fix and a question about symbolic gradient.

mruberry reviewed (approved):
Cool! Thanks @BIT-silence! Really appreciate you taking the time to make the edge case behavior consistent and to validate it!

xiaomengy committed:
Add torch.logit function (#41062)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41062

Add torch.logit function

Test Plan: buck test mode/dev-nosan //caffe2/test:torch -- "logit"

Differential Revision: D22406912

fbshipit-source-id: 0c66ce78b2ae82dfbe39a221f857c0330a7acb86
